In [1]:
import pandas as pd
import numpy as np

### Q1. Pandas version
What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.3.2'

### Getting the data
For this homework, we'll use the Car Fuel Efficiency dataset.
You can do it with wget:

```wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv```

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [4]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

--2025-09-17 15:15:31--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv’

car_fuel_efficiency 100%[===================>] 853.70K  4.80MB/s    in 0.2s    

2025-09-17 15:15:31 (4.80 MB/s) - ‘car_fuel_efficiency.csv’ saved [874188/874188]



In [3]:
!ls

01.ipynb  car_fuel_efficiency.csv


In [4]:
df = pd.read_csv("car_fuel_efficiency.csv")

### Q2. Records count
How many records are in the dataset?

In [5]:
df.shape

(9704, 11)

Ans. 9704

### Q3. Fuel types
How many fuel types are presented in the dataset?

In [6]:
df["fuel_type"].unique()

array(['Gasoline', 'Diesel'], dtype=object)

Ans. 2

### Q4. Missing values
How many columns in the dataset have missing values?

In [7]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

Ans. 4

### Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [8]:
print(df.loc[df["origin"] == "Asia", "fuel_efficiency_mpg"].max())

23.759122836520497


Ans. 23.75

### Q6. Median value of horsepower
1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

Has it changed?

In [9]:
df["horsepower"].describe()

count    8996.000000
mean      149.657292
std        29.879555
min        37.000000
25%       130.000000
50%       149.000000
75%       170.000000
max       271.000000
Name: horsepower, dtype: float64

In [10]:
df["horsepower"].mode()

0    152.0
Name: horsepower, dtype: float64

In [11]:
horsepower_mode = df["horsepower"].fillna(df["horsepower"].mode())

In [12]:
horsepower_mode.describe()

count    8996.000000
mean      149.657292
std        29.879555
min        37.000000
25%       130.000000
50%       149.000000
75%       170.000000
max       271.000000
Name: horsepower, dtype: float64

Ans. No

### Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the sum of all the elements of the result?

In [13]:
X = df.loc[df["origin"] == "Asia", ["vehicle_weight", "model_year"]].head(7).to_numpy()

In [14]:
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [15]:
X.dtype

dtype('float64')

In [16]:
XTX_inv = np.linalg.inv(np.dot(X.T, X))

In [17]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [18]:
w_n = np.dot(XTX_inv, X.T)
w = np.dot(w_n, y)
print(w)

[0.01386421 0.5049067 ]


In [19]:
print(np.sum(w))

0.5187709081074007


Ans. 0.51